### PR TITLE
Rule 6.1.5: Explicit supertype qualification should not be used if there is not clash between called methods

### LINKED ISSUE #441: Rule 6.1.5: Explicit supertype qualification should not be used if there is not clash between called methods
This rule is applicable for both interfaces and classes.

Bad example:
```kotlin
open class Rectangle {
    open fun draw() { /* ... */ }
}

class Square() : Rectangle() {
    override fun draw() {
        super<Rectangle>.draw() // no need in super<Rectangle> here
    }
}
```

### PR DESCRIPTION
**Pull request by kentr0w**
### What's done:
   Implemented rule 6.1.5

## Actions checklist
* [x] Implemented Rule, added Warnings
* [x] Added tests on checks
* [x] Added tests on fixers
* [x] Updated diktat-analysis.yml
* [x] Updated available-rules.md
* [x] closes #441

### CODE REVIEW

**Review by petertrr — COMMENTED** [review]
You are checking not quite what is written in the rule. If it has been discussed, please update rule text.

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/UselessSupertype.kt [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+        emitWarn = emit
+        isFixMode = autoCorrect
+
+        if (node.elementType == CLASS)
+            checkClass(node)
+    }
+
+    private fun checkClass(node: ASTNode) {
+        val superNodes = node.findAllNodesWithCondition({ it.elementType in SUPER_TYPE }).ifEmpty { return }
+        val overrideNodes = node.findAllNodesWithSpecificType(DOT_QUALIFIED_EXPRESSION)
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/UselessSupertype.kt:51]:
```suggestion
        val qualifiedSuperCalls = node.findAllNodesWithSpecificType(DOT_QUALIFIED_EXPRESSION)
```
I don't think these nodes override anything

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/UselessSupertype.kt [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+                .mapNotNull { findFunWithSuper(it) }.ifEmpty { return }
+        if (superNodes.size == 1)
+            overrideNodes.map { removeSupertype(it.first) }
+        else
+            handleManyImpl(superNodes, overrideNodes)
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/UselessSupertype.kt:56]:
```suggestion
        if (superNodes.size == 1) {
            overrideNodes.map { removeSupertype(it.first) }
        } else {
            handleManyImpl(superNodes, overrideNodes)
        }
```
there is a rule in diktat that forces to always use braces :)

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/DiktatRuleSetProvider.kt [hunk 3]
```diff
@@ -62,6 +62,8 @@ class DiktatRuleSetProvider(private val diktatConfigFile: String = "diktat-analy
                 ::PackageNaming,
                 ::IdentifierNaming,
                 // code structure
+                ::UselessSupertype,
+                ::LocalVariablesRule,
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/DiktatRuleSetProvider.kt:66]: Did you move `LocalVariablesRule` on purpose?

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/constants/Warnings.kt [hunk 4]
```diff
@@ -121,6 +121,7 @@ enum class Warnings(private val canBeAutoCorrected: Boolean, private val warn: S
     WRONG_NAME_OF_VARIABLE_INSIDE_ACCESSOR(false, "Use `field` keyword instead of property name inside property accessors"),
     MULTIPLE_INIT_BLOCKS(true, "Avoid using multiple `init` blocks, this logic can be moved to constructors or properties declarations"),
     CLASS_SHOULD_NOT_BE_ABSTRACT(true, "class should not be abstract, because it has no abstract functions"),
+    USELESS_OVERRIDE(true,"override method doesn't differ from super"),
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/constants/Warnings.kt:124]: I think warning name and description should be changed.

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/UselessSupertype.kt [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+import com.pinterest.ktlint.core.ast.parent
+import org.cqfn.diktat.common.config.rules.RulesConfig
+import org.cqfn.diktat.ruleset.constants.Warnings.USELESS_OVERRIDE
+import org.cqfn.diktat.ruleset.utils.*
+import org.jetbrains.kotlin.com.intellij.lang.ASTNode
+import org.jetbrains.kotlin.psi.psiUtil.siblings
+
+/**
+ * rule 6.1.5
+ * Explicit supertype qualification should not be used if there is not clash between called methods
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/UselessSupertype.kt:28]: Please add fixme comment that we can't analyze supertypes that are defined in other files.

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/UselessSupertype.kt [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+     * The method looks in the same file for all super interfaces or a class, in each it looks for methods
+     * that can be overridden and creates a map with a key - the name of the method and value - the number of times it meets
+     *
+     * @param superTypeList - list of identifiers super classes
+     * @param methodsName - name of overrides methods
+     *
+     * @return map name of method and the number of times it meets
+     */
+    @Suppress("UnsafeCallOnNullableType")
+    private fun findAllSupers(superTypeList: List<ASTNode>, methodsName: List<String>): MutableMap<String, Int>? {
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/UselessSupertype.kt:110]: Why `MutableMap`?

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/UselessSupertype.kt [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+                                it.getIdentifierName()!!.text in methodsName
+                    }
+            overrideFunctions.forEach {
+                if (functionNameMap.containsKey(it.getIdentifierName()!!.text))
+                    functionNameMap[it.getIdentifierName()!!.text] = functionNameMap[it.getIdentifierName()!!.text]!! + 1
+                else
+                    functionNameMap[it.getIdentifierName()!!.text] = 1
+            }
+        }
+        return functionNameMap
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/UselessSupertype.kt:133]:
```suggestion
        return functionNameMap.toMap()
```
I believe this way you can convert mutable map to immutable, assuming you won't need to change it afterwards.

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/UselessSupertype.kt [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+            val overrideFunctions = classBody.findAllNodesWithSpecificType(FUN)
+                    .filter {
+                        (if (classBody.treeParent.hasChildOfType(CLASS_KEYWORD)) it.findChildByType(MODIFIER_LIST)!!.hasChildOfType(OPEN_KEYWORD) else true) &&
+                                it.getIdentifierName()!!.text in methodsName
+                    }
+            overrideFunctions.forEach {
+                if (functionNameMap.containsKey(it.getIdentifierName()!!.text))
+                    functionNameMap[it.getIdentifierName()!!.text] = functionNameMap[it.getIdentifierName()!!.text]!! + 1
+                else
+                    functionNameMap[it.getIdentifierName()!!.text] = 1
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/UselessSupertype.kt:130]: Take a look at `MutableMap#compute` method, it should let you rewrite this whole `if`.

**Review by petertrr — CHANGES_REQUESTED** [review]

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/DiktatRuleSetProvider.kt (hunk 3 above)
Comment (kentr0w) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/DiktatRuleSetProvider.kt:66]: No, it was my a mistake during the merge

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/UselessSupertype.kt [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+     *
+     * @return pair of identifier
+     */
+    @Suppress("UnsafeCallOnNullableType")
+    private fun findFunWithSuper(node: ASTNode): Pair<ASTNode, ASTNode>? {
+        return Pair(
+                node.findChildByType(SUPER_EXPRESSION)
+                        ?.findChildByType(TYPE_REFERENCE)
+                        ?.findAllNodesWithSpecificType(IDENTIFIER)
+                        ?.first() ?: return null,
```
[the remark this answers is not in the document]
Comment (kentr0w) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/UselessSupertype.kt:94]: Yes you are right. I checked

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/UselessSupertype.kt [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+                }
+    }
+
+    @Suppress("UnsafeCallOnNullableType")
+    private fun removeSupertype(node: ASTNode) {
+        USELESS_SUPERTYPE.warnAndFix(configRules, emitWarn, isFixMode, node.text, node.startOffset, node) {
+            val startNode = node.parent({ it.elementType == SUPER_EXPRESSION })!!.findChildByType(REFERENCE_EXPRESSION)!!
+            val lastNode = startNode.siblings(true).last()
+            startNode.treeParent.removeRange(startNode.treeNext, lastNode)
+            startNode.treeParent.removeChild(lastNode)
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/UselessSupertype.kt:79]: So, here you remove all children of `SUPER_EXPRESSION`. But the `SUPER_EXPRESSION` node itself is left in the tree? You should remove it explicitly too, even though it's empty.

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/UselessSupertype.kt [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+                .mapNotNull { findFunWithSuper(it) }.ifEmpty { return }
+        if (superNodes.size == 1) {
+            qualifiedSuperCalls.map { removeSupertype(it.first) }
+        } else {
+            handleManyImpl(superNodes, qualifiedSuperCalls)
+        }
+    }
+
+    private fun handleManyImpl(superNodes: List<ASTNode>, overrideNodes: List<Pair<ASTNode, ASTNode>>) {
+        val uselessSuperType = findAllSupers(superNodes, overrideNodes.map { it.second.text })?.filter { it.value == 1 }?.map { it.key }
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/UselessSupertype.kt:63]:
```suggestion
        val uselessSuperType = findAllSupers(superNodes, overrideNodes.map { it.second.text })
            ?.filter { it.value == 1 }
            ?.map { it.key }
```
This line looks a bit like magic :) Please add some comments.

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/UselessSupertype.kt [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+        val functionNameMap = mutableMapOf<String, Int>()
+        superNodes.forEach { classBody ->
+            val overrideFunctions = classBody.findAllNodesWithSpecificType(FUN)
+                    .filter {
+                        (if (classBody.treeParent.hasChildOfType(CLASS_KEYWORD)) it.findChildByType(MODIFIER_LIST)!!.hasChildOfType(OPEN_KEYWORD) else true) &&
+                                it.getIdentifierName()!!.text in methodsName
+                    }
+            overrideFunctions.forEach {
+                functionNameMap[it.getIdentifierName()!!.text] = functionNameMap.getOrDefault(it.getIdentifierName()!!.text, 0) + 1
+                /*if (functionNameMap.containsKey(it.getIdentifierName()!!.text))
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/UselessSupertype.kt:131]: Commented out code?

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/UselessSupertype.kt [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+            return null
+        val functionNameMap = mutableMapOf<String, Int>()
+        superNodes.forEach { classBody ->
+            val overrideFunctions = classBody.findAllNodesWithSpecificType(FUN)
+                    .filter {
+                        (if (classBody.treeParent.hasChildOfType(CLASS_KEYWORD)) it.findChildByType(MODIFIER_LIST)!!.hasChildOfType(OPEN_KEYWORD) else true) &&
+                                it.getIdentifierName()!!.text in methodsName
+                    }
+            overrideFunctions.forEach {
+                functionNameMap[it.getIdentifierName()!!.text] = functionNameMap.getOrDefault(it.getIdentifierName()!!.text, 0) + 1
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/UselessSupertype.kt:130]: I think you could use `compute` method like `functionNameMap.compute(idntifier) { key, oldValue -> ... }` where `oldValue` will be `null` if the `key` is not present in the map yet.

**Review by petertrr — CHANGES_REQUESTED** [review]
Looks more or less fine, some more suggestions about documentation and style.

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/UselessSupertype.kt (hunk 10 above)
Comment (kentr0w) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/UselessSupertype.kt:79]: No, I shouldn't remove `SUPER_EXPRESSION`.

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/UselessSupertype.kt [hunk 14]
```diff
@@ -0,0 +1,135 @@
...
+    @Suppress("UnsafeCallOnNullableType")
+    private fun findFunWithSuper(node: ASTNode): Pair<ASTNode, ASTNode>? {
+        return Pair(
+                node.findChildByType(SUPER_EXPRESSION)
+                        ?.findChildByType(TYPE_REFERENCE)
+                        ?.findAllNodesWithSpecificType(IDENTIFIER)
+                        ?.first() ?: return null,
+                node.findChildByType(CALL_EXPRESSION)
+                        ?.findAllNodesWithSpecificType(IDENTIFIER)
+                        ?.first() ?: return null)
```
[the remark this answers is not in the document]
Comment (petertrr) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/UselessSupertype.kt]: Yes, we should check selector expression whether it is a function or a property.

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/DiktatRuleSetProvider.kt (hunk 3 above)
Comment (petertrr) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/DiktatRuleSetProvider.kt:68]: Please move `LocalVariablesRule` back if it's unintended move, the rest seems fine.

**Review by petertrr — CHANGES_REQUESTED** [review]

**Review by petertrr — APPROVED** [review]

**Pull request merged**